arangod/Aql/QueryList.cpp

```diff
@@ -262,85 +265,93 @@ void QueryList::remove(Query& query) {
 
 /// @brief kills a query
 Result QueryList::kill(TRI_voc_tick_t id) {
-  size_t const maxLength =
-      _maxQueryStringLength.load(std::memory_order_relaxed);
-
-  READ_LOCKER(writeLocker, _lock);
-
-  auto it = _current.find(id);
-
-  if (it == _current.end()) {
-    return {TRI_ERROR_QUERY_NOT_FOUND, "query ID not found in query list"};
+  // We should call dtor std::shared_ptr<Query> not under a lock,
+  // otherwise a deadlock will occur. Query dtor make a WRITE_LOCKER(..., _lock)
+  std::shared_ptr<Query> queryPtr;
+  {
+    READ_LOCKER(writeLocker, _lock);
+    auto it = _current.find(id);
+    if (it == _current.end()) {
+      return {TRI_ERROR_QUERY_NOT_FOUND, "query ID not found in query list"};
+    }
+    queryPtr = it->second.lock();
   }
-
-  killQuery(*it->second, maxLength, false);
-
-  return Result();
+  if (queryPtr) {
+    auto const length = _maxQueryStringLength.load(std::memory_order_relaxed);
+    killQuery(*queryPtr, length, false);
+  }
+  return {};
 }
 
 /// @brief kills all currently running queries that match the filter function
 /// (i.e. the filter should return true for a queries to be killed)
 uint64_t QueryList::kill(std::function<bool(Query&)> const& filter,
                          bool silent) {
+  // We should call dtor std::shared_ptr<Query> not under a lock,
+  // otherwise a deadlock will occur. Query dtor make a WRITE_LOCKER(..., _lock)
+  absl::InlinedVector<std::shared_ptr<Query>, 16> queries;
+  {
+    READ_LOCKER(readLocker, _lock);
+    queries.reserve(_current.size());
+    for (auto& it : _current) {
+      if (auto queryPtr = it.second.lock(); queryPtr) {
+        queries.push_back(std::move(queryPtr));
+      }
+    }
+  }
   uint64_t killed = 0;
-  size_t const maxLength =
-      _maxQueryStringLength.load(std::memory_order_relaxed);
-
-  READ_LOCKER(readLocker, _lock);
-
-  for (auto& it : _current) {
-    Query& query = *(it.second);
-
-    if (!filter(query)) {
-      continue;
+  auto const maxLength = _maxQueryStringLength.load(std::memory_order_relaxed);
+  for (auto const& queryPtr : queries) {
+    if (auto& query = *queryPtr; filter(query)) {
+      killQuery(query, maxLength, silent);
+      ++killed;
     }
-
-    killQuery(query, maxLength, silent);
-    ++killed;
   }
-
   return killed;
 }
 
 /// @brief get the list of currently running queries
 std::vector<QueryEntryCopy> QueryList::listCurrent() {
+  // We should call dtor std::shared_ptr<Query> not under a lock,
+  // otherwise a deadlock will occur. Query dtor make a WRITE_LOCKER(..., _lock)
```

Comment:
```suggestion
  // We must not call the std::shared_ptr<Query> dtor under the lock,
  // because this can result in a deadlock since the Query dtor tries to
  // remove itself from this list which acquires the `_lock` (again).
```

Reply:
I agree, this sounds better, but I think the important point is that in dtor we are trying to acquire a write lock.